Snakefile

```diff
@@ -0,0 +1,402 @@
+import subprocess
+
+#################################################################
+####################### general setup ###########################
+#################################################################
+
+# certain files can be fetched from the database or downloaded from our S3 bucket.
+# config["fetch_from_database"] dictates this
+# We could implement this via remote snakemake inputs (such as in nextstrain/ncov),
+# conditional rules (a bit of an anti-pattern) or ruleorders.
+ruleorder: download_previous_biosample > fetch_biosample_from_database
+ruleorder: download_previous_raw_ndjson > fetch_from_database
+if config.get("fetch_from_database", False):
+    ruleorder: fetch_biosample_from_database > download_previous_biosample
+    ruleorder: fetch_from_database > download_previous_raw_ndjson
+
+send_notifications = "SLACK_CHANNELS" in os.environ and "SLACK_TOKEN" in os.environ
+
+#################################################################
+################ work out what steps to run #####################
+#################################################################
+
+def _compute_targets(wildcards):
+    database = config["database_name"]    
+    targets = [f"data/{database}/upload.done"]
+    if config.get("trigger_preprocessing", False):
+        targets.append(f"data/{database}/trigger-preprocessing.done")
+    if send_notifications:
+        targets.append(f"data/{database}/notify.done")
+    if config.get("fetch_from_database", False):
+        targets.append(f"data/{database}/raw.upload.done")
+    return targets
+
+rule all:
+    input: _compute_targets
+
+#################################################################
+###################### rule definitions #########################
+#################################################################
+
+
+def run_shell_command_n_times(cmd, msg, cleanup_failed_cmd, retry_num=5):
+    attempt = 0
+    while attempt < retry_num:
+        print(f"{msg} attempt number {attempt}")
+        try:
+            subprocess.run(cmd, shell=True, check=True)
```

Comment:
Agree. This should probably happen before merge, so we don't get bitten sooner or later by this issue.

Reply:
Yes it can — it turns out that when `shell()` fails it throws a `subprocess.CalledProcessError` so we can even use  the same try/except logic here! Implemented in ab59826